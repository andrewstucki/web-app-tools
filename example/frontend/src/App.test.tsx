import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { User } from "./models";

const injectState = (mockUser: User | null) => {
  const mockDispatch = jest.fn();
  jest.mock("react-redux", () => ({
    ...jest.requireActual("react-redux"),
    useSelector: (): User | null => mockUser,
    useDispatch: () => mockDispatch,
  }));
  const App = require("./App").App;
  return { App, dispatch: mockDispatch };
};

describe("App", () => {
  beforeEach(() => {
    jest.resetModules();
  });

  it("renders loading state with no user", () => {
    const { App } = injectState(null);
    expect(toJson(shallow(<App />))).toMatchSnapshot();
  });

  it("requests the user's profile", () => {
    const { App, dispatch } = injectState(null);
    const node = toJson(shallow(<App />));
    expect(dispatch).toBeCalledWith(
      expect.objectContaining({
        type: "__AUTHENTICATED_REQUEST",
        payload: expect.objectContaining({
          config: { url: "/api/v1/me" },
        }),
      })
    );
  });

  it("renders basic info given a user", () => {
    const { App } = injectState({
      id: "foo",
      email: "bar@baz.com",
      createdAt: new Date(),
      updatedAt: new Date(),
    });
    expect(toJson(shallow(<App />))).toMatchSnapshot();
  });

  it("logs out properly", () => {
    const { App, dispatch } = injectState({
      id: "foo",
      email: "bar@baz.com",
      createdAt: new Date(),
      updatedAt: new Date(),
    });
    const node = shallow(<App />)
      .find("button")
      .simulate("click");
    expect(dispatch).toBeCalledWith(
      expect.objectContaining({
        type: "__AUTHENTICATED_LOG_OUT",
      })
    );
  });
});
