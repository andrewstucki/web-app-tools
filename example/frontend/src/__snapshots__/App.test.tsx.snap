// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders basic info given a user 1`] = `
<div>
  <h1>
    Logged in as:
  </h1>
  <table>
    <tr>
      <td>
        ID
      </td>
      <td>
        foo
      </td>
    </tr>
    <tr>
      <td>
        Email
      </td>
      <td>
        bar@baz.com
      </td>
    </tr>
    <tr>
      <td>
        Created At
      </td>
      <td>
        a few seconds ago
      </td>
    </tr>
    <tr>
      <td>
        Last Updated
      </td>
      <td>
        a few seconds ago
      </td>
    </tr>
  </table>
  <button
    onClick={[Function]}
  >
    Log Out
  </button>
</div>
`;

exports[`App renders loading state with no user 1`] = `
<div>
  Loading
</div>
`;
